File: PFinalQytetetJava/QytetetJava/src/controladorqytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package controladorqytetet;

import modeloqytetet.EstadoJuego;
import modeloqytetet.Qytetet;
import modeloqytetet.MetodoSalirCarcel;

import java.util.ArrayList;
import modeloqytetet.Jugador;

/**
 *
 * @author antonio-w10
 */
public class ControladorQytetet {
    
    
    private ArrayList<String> nombreJugadores;
    
    private Qytetet modelo = Qytetet.getInstance();
    
    private ControladorQytetet() {
    }
    
    
    private static final ControladorQytetet instance = new ControladorQytetet();
    
    public static ControladorQytetet getInstance() {
        return instance;
    }
    
    
    
    public void setNombreJugadores( ArrayList<String> nombreJugadores){
        this.nombreJugadores = nombreJugadores;
    }
    
    public ArrayList<Integer> obtenerOperacionesJuegoValidas() {
        
        ArrayList<Integer> operaciones= new ArrayList<>();
                
        if (modelo.getJugadores().isEmpty()){
            operaciones.add(OpcionMenu.INICIARJUEGO.ordinal());
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_CONSORPRESA){
            operaciones.add(OpcionMenu.APLICARSORPRESA.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.ALGUNJUGADORENBANCARROTA){            
            operaciones.add(OpcionMenu.OBTENERRANKING.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_PUEDECOMPRAROGESTIONAR){           
            operaciones.add(OpcionMenu.PASARTURNO.ordinal());
            operaciones.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
            operaciones.add(OpcionMenu.EDIFICARCASA.ordinal());
            operaciones.add(OpcionMenu.EDIFICARHOTEL.ordinal());
            operaciones.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
            operaciones.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
            operaciones.add(OpcionMenu.COMPRARTITULOPROPIEDAD.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
            
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_PUEDEGESTIONAR){            
            operaciones.add(OpcionMenu.PASARTURNO.ordinal());
            operaciones.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
            operaciones.add(OpcionMenu.EDIFICARCASA.ordinal());
            operaciones.add(OpcionMenu.EDIFICARHOTEL.ordinal());
            operaciones.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
            operaciones.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_PREPARADO){            
            operaciones.add(OpcionMenu.JUGAR.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_ENCARCELADO){            
            operaciones.add(OpcionMenu.PASARTURNO.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        }
        else if (modelo.getEstadoJuego() == EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD){            
            operaciones.add(OpcionMenu.INTENTARSALIRCARCELPAGANDOLIBERTAD.ordinal());
            operaciones.add(OpcionMenu.INTENTARSALIRCARCELTIRANDODADO.ordinal());
            
            operaciones.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
            operaciones.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
            operaciones.add(OpcionMenu.MOSTRARTABLERO.ordinal());
            operaciones.add(OpcionMenu.TERMINARJUEGO.ordinal());
        }
        
        
        
        
        
    
        return operaciones;
    }
    public boolean necesitaElegirCasilla(int opcionMenu) {
        
        boolean puede = false;
        if (opcionMenu == OpcionMenu.HIPOTECARPROPIEDAD.ordinal()
                || opcionMenu == OpcionMenu.CANCELARHIPOTECA.ordinal()
                || opcionMenu == OpcionMenu.EDIFICARCASA.ordinal()
                || opcionMenu == OpcionMenu.EDIFICARHOTEL.ordinal()
                || opcionMenu == OpcionMenu.VENDERPROPIEDAD.ordinal() ) {
            
            
            puede = true;
        }
        
        
        return puede;
    }
    
    public ArrayList<Integer> obtenerCasillasValidas(int opcionMenu) {
        
        ArrayList<Integer> operaciones = new ArrayList<>();
        
         if(opcionMenu == OpcionMenu.HIPOTECARPROPIEDAD.ordinal()){
            operaciones = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
        }
        
        else if(opcionMenu == OpcionMenu.CANCELARHIPOTECA.ordinal()){
            operaciones = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true);
        }
        else{
            operaciones = modelo.obtenerPropiedadesJugador();
        }
               
        return operaciones;
    }
    
    public String realizarOperacion(int opcionElegida, int casillaElegida) {
        
        String mensaje = "naaaada";
        
        if (opcionElegida == OpcionMenu.INICIARJUEGO.ordinal()) {
            
            mensaje = "Juego Iniciado";
            modelo.inicializarJuego(nombreJugadores);
            
        }
        
        if (opcionElegida == OpcionMenu.JUGAR.ordinal()) {
            
            mensaje = "JUGANDO";
            modelo.jugar();
            
        }
        
        if (opcionElegida == OpcionMenu.APLICARSORPRESA.ordinal()) {
            
            modelo.aplicarSorpresa();
            mensaje = "SORPRESA APLICADA";
        }
        
        if (opcionElegida == OpcionMenu.INTENTARSALIRCARCELPAGANDOLIBERTAD.ordinal()) {
            
            
            if(modelo.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD)){
                mensaje = "SALIR CARCEL CORRECTO";
            }
            else{
                mensaje = "SALIR CARCEL  FALLIDO";
            }
        }
        
        if (opcionElegida == OpcionMenu.INTENTARSALIRCARCELTIRANDODADO.ordinal()) {
            
            if(modelo.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO)){
                mensaje = "SALIR CARCEL CORRECTO";
            }
            else{
                mensaje = "SALIR CARCEL  FALLIDO";
            }
            

        }
        
        if (opcionElegida == OpcionMenu.COMPRARTITULOPROPIEDAD.ordinal()) {
            
            if(modelo.comprarTituloPropiedad()){
                mensaje = "COMPRAR CORRECTO";
            }
            else{
                mensaje = "COMPRAR FALLIDO";
            }
            
            
        }
        
        if (opcionElegida == OpcionMenu.HIPOTECARPROPIEDAD.ordinal()) {
            
            modelo.hipotecarPropiedad(casillaElegida);
            mensaje = "PROPIEDAD HIPOTECADA";
        }
        
        if (opcionElegida == OpcionMenu.CANCELARHIPOTECA.ordinal()) {
            
            
            if(modelo.cancelarHipoteca(casillaElegida)){
                mensaje = "CANCELAR HIPOTECA CORRECTO";
            }
            else{
                mensaje = "CANCELAR HIPOTECA FALLIDO";
            }
        }
        
        if (opcionElegida == OpcionMenu.EDIFICARCASA.ordinal()) {
            
            if(modelo.edificarCasa(casillaElegida)){
                mensaje = "EDIFICAR CASA CORRECTO";
            }
            else{
                mensaje = "EDIFICAR CASA FALLIDO";
            }
        }
        
        if (opcionElegida == OpcionMenu.EDIFICARHOTEL.ordinal()) {
                        
            if(modelo.edificarHotel(casillaElegida)){
                mensaje = "EDIFICAR HOTEL CORRECTO";
            }
            else{
                mensaje = "EDIFICAR HOTEL FALLIDO";
            }
        }
        
        if (opcionElegida == OpcionMenu.VENDERPROPIEDAD.ordinal()) {
            
            modelo.venderPropiedad(casillaElegida);
            mensaje = "PROPIEDAD VENDIDA";
        }
        
        if (opcionElegida == OpcionMenu.PASARTURNO.ordinal()) {
            
            modelo.siguienteJugador();
            mensaje = "TURNO del SIGUIENTE JUGADOR";
        }
        
        if (opcionElegida == OpcionMenu.OBTENERRANKING.ordinal()) {
            mensaje = "realizar operacion";
            modelo.obtenerRanking();
            
        }
        
        if (opcionElegida == OpcionMenu.TERMINARJUEGO.ordinal()) {
            
            System.exit(0);
        }
        
        if (opcionElegida == OpcionMenu.MOSTRARJUGADORACTUAL.ordinal()) {
            mensaje = "realizar operacion";
            System.out.print(modelo.getJugadorActual().toString());
            
        }
        
        if (opcionElegida == OpcionMenu.MOSTRARJUGADORES.ordinal()) {
            mensaje = "realizar operacion";
            for (Jugador jug: modelo.getJugadores()){                
               System.out.println(jug.toString());
            }  
            
            
        }
        if (opcionElegida == OpcionMenu.MOSTRARTABLERO.ordinal()) {
            mensaje = "realizar operacion";
            System.out.print(modelo.getTablero().toString());
            
        }
        
        return mensaje;
    }

   
    
}

File: PFinalQytetetRuby/QytetetRuby/lib/controlador_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#require_relative "@modeloQytetet"
require_relative "qytetet"
require_relative "metodo_salir_carcel"
require_relative "estado_juego"
require_relative "opcion_menu"


require "singleton"


module Controladorqytetet
  class ControladorQytetet
    
    
    include Singleton
    include ModeloQytetet
    
    attr_accessor :nombreJugadores
    
    @@modelo = Qytetet.instance
    
    def initialize
      @nombreJugadores = Array.new
      
    end
    
    
    
   def obtenerOperacionesJuegoValidas() 
        
        operaciones = Array.new
                
        if (@@modelo.jugadores.empty?())
            operaciones << (OpcionMenu.index(:INICIARJUEGO))
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_CONSORPRESA)
            operaciones << (OpcionMenu.index(:APLICARSORPRESA))            
       
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        
        elsif (@@modelo.estadoJuego == EstadoJuego::ALGUNJUGADORENBANCARROTA)         
            operaciones << (OpcionMenu.index(:OBTENERRANKING))
            
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_PUEDECOMPRAROGESTIONAR)
            operaciones << (OpcionMenu.index(:PASARTURNO))
            operaciones << (OpcionMenu.index(:VENDERPROPIEDAD))
            operaciones << (OpcionMenu.index(:EDIFICARCASA))
            operaciones << (OpcionMenu.index(:EDIFICARHOTEL))
            operaciones << (OpcionMenu.index(:HIPOTECARPROPIEDAD))
            operaciones << (OpcionMenu.index(:CANCELARHIPOTECA))
            operaciones << (OpcionMenu.index(:COMPRARTITULOPROPIEDAD))
       
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
            
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_PUEDEGESTIONAR)        
            operaciones << (OpcionMenu.index(:PASARTURNO))
            operaciones << (OpcionMenu.index(:VENDERPROPIEDAD))
            operaciones << (OpcionMenu.index(:EDIFICARCASA))
            operaciones << (OpcionMenu.index(:EDIFICARHOTEL))
            operaciones << (OpcionMenu.index(:HIPOTECARPROPIEDAD))
            operaciones << (OpcionMenu.index(:CANCELARHIPOTECA))
            
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_PREPARADO)          
            
            operaciones << (OpcionMenu.index(:JUGAR))
       
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_ENCARCELADO)     
            
            operaciones << (OpcionMenu.index(:PASARTURNO))
       
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        
        elsif (@@modelo.estadoJuego == EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD)        
            
            operaciones << (OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD))
            operaciones << (OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO))
       
       
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORACTUAL))
            operaciones << (OpcionMenu.index(:MOSTRARJUGADORES))
            operaciones << (OpcionMenu.index(:MOSTRARTABLERO))
            operaciones << (OpcionMenu.index(:TERMINARJUEGO))
        end
        
        
        
        
        
    
        return operaciones
    end
    
    def necesitaElegirCasilla(opcionMenu) 
        
        puede = false
        if (opcionMenu == (OpcionMenu.index(:VENDERPROPIEDAD)) || opcionMenu == (OpcionMenu.index(:EDIFICARCASA)) || opcionMenu == (OpcionMenu.index(:EDIFICARHOTEL)) || opcionMenu == (OpcionMenu.index(:HIPOTECARPROPIEDAD)) || opcionMenu == (OpcionMenu.index(:CANCELARHIPOTECA)))
            
            puede = true
        end
        
        
        return puede
    end
    
    def obtenerCasillasValidas(opcionMenu)
        
        operaciones = Array.new
        
         if(opcionMenu == (OpcionMenu.index(:HIPOTECARPROPIEDAD)))
            operaciones = @@modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false)
         
        
        elsif(opcionMenu == (OpcionMenu.index(:CANCELARHIPOTECA)))
            operaciones = @@modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true)
        
        else
            operaciones = @@modelo.obtenerPropiedadesJugador()
        end
               
        return operaciones
    end
    
    def realizarOperacion(opcionElegida, casillaElegida) 
        
        mensaje = "naaaada"
        
        if (opcionElegida == (OpcionMenu.index(:INICIARJUEGO)))
            
            mensaje = "Juego Iniciado"
            @@modelo.inicializarJuego(@nombreJugadores)
            
        end
        
        if (opcionElegida == (OpcionMenu.index(:JUGAR)))
            
            mensaje = "JUGANDO"
            @@modelo.jugar()
            
        end
        
        if (opcionElegida == (OpcionMenu.index(:APLICARSORPRESA)))
            
            @@modelo.aplicarSorpresa()
            mensaje = "SORPRESA APLICADA"
        end
        
        if (opcionElegida == (OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD)))
            
            
            if(@@modelo.intentarSalirCarcel(MetodoSalirCarcel::PAGANDOLIBERTAD))
                mensaje = "SALIR CARCEL CORRECTO"
            
            else
                mensaje = "SALIR CARCEL  FALLIDO"
            end
        end
        
        if (opcionElegida == (OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO)))
            
            if(@@modelo.intentarSalirCarcel(MetodoSalirCarcel::TIRANDODADO))
                mensaje = "SALIR CARCEL CORRECTO"
            
            else
                mensaje = "SALIR CARCEL  FALLIDO"
            end
            

        end
        
        if (opcionElegida == (OpcionMenu.index(:COMPRARTITULOPROPIEDAD)))
            
            if(@@modelo.comprarTituloPropiedad())
                mensaje = "COMPRAR CORRECTO"
            
            else
                mensaje = "COMPRAR FALLIDO"
            end
            
            
        end
        
        if (opcionElegida == (OpcionMenu.index(:HIPOTECARPROPIEDAD))) 
            
            @@modelo.hipotecarPropiedad(casillaElegida)
            mensaje = "PROPIEDAD HIPOTECADA"
        end
        
        if (opcionElegida == (OpcionMenu.index(:CANCELARHIPOTECA)))
            
            
            if(@@modelo.cancelarHipoteca(casillaElegida))
                mensaje = "CANCELAR HIPOTECA CORRECTO"
            
            else
                mensaje = "CANCELAR HIPOTECA FALLIDO"
            end
        end
        
        if (opcionElegida == (OpcionMenu.index(:EDIFICARCASA)))
            
            if(@@modelo.edificarCasa(casillaElegida))
                mensaje = "EDIFICAR CASA CORRECTO"
            
            else
                mensaje = "EDIFICAR CASA FALLIDO"
            end
        end
        
        if (opcionElegida == (OpcionMenu.index(:EDIFICARHOTEL)))
                        
            if(@@modelo.edificarHotel(casillaElegida))
                mensaje = "EDIFICAR HOTEL CORRECTO"
            
            else
                mensaje = "EDIFICAR HOTEL FALLIDO"
            end
        end
        
        if (opcionElegida == (OpcionMenu.index(:VENDERPROPIEDAD)))
            
            @@modelo.venderPropiedad(casillaElegida)
            mensaje = "PROPIEDAD VENDIDA"
        end
        
        if (opcionElegida == (OpcionMenu.index(:PASARTURNO)))
            
            @@modelo.siguienteJugador()
            mensaje = "TURNO del SIGUIENTE JUGADOR"
        end
        
        if (opcionElegida == (OpcionMenu.index(:OBTENERRANKING)))
            mensaje = "realizada operacion"
            @@modelo.obtenerRanking()
            
        end
        
        if (opcionElegida == (OpcionMenu.index(:TERMINARJUEGO)))
            
            exit
        end
        
        if (opcionElegida == (OpcionMenu.index(:MOSTRARJUGADORACTUAL)))
            mensaje = "realizada operacion"
            puts (@@modelo.jugadorActual.to_s)
            
        end
        
        if (opcionElegida == (OpcionMenu.index(:MOSTRARJUGADORES)))
            mensaje = "realizada operacion"
            @@modelo.jugadores.each do |jug|      
              puts (jug.to_s)
              puts
            end     
      
            
        end
        if (opcionElegida == (OpcionMenu.index(:MOSTRARTABLERO)))
            mensaje = "realizada operacion"
            puts (@@modelo.tablero.to_s)
            
        end
        
        return mensaje
    end
    
    
    
    
    
  end
end
